synapse/handlers/cas_handler.py

```diff
@@ -127,31 +154,46 @@ def _parse_cas_response(
             )
         return user, attributes
 
-    async def _on_successful_auth(
-        self,
-        username: str,
-        request: SynapseRequest,
-        client_redirect_url: str,
-        user_display_name: Optional[str] = None,
+    def get_redirect_url(self, service_redirect_endpoint: str, **kwargs) -> str:
+        """
+        Generates a URL to the CAS server where the client should be redirected.
+
+        Args:
+            service_redirect_endpoint: The homeserver endpoint to redirect
+                the client to after successful SSO negotiation.
+            kwargs: Additional arguments to include in the final redirect URL.
+
+        Returns:
+            The URL to redirect the client to.
+        """
+        args = urllib.parse.urlencode(
+            {"service": self._build_service_param(service_redirect_endpoint, **kwargs)}
+        )
+
+        return "%s/login?%s" % (self._cas_server_url, args)
+
+    async def handle_login_request(
+        self, request: SynapseRequest, client_redirect_url: str, ticket: str
     ) -> None:
-        """Called once the user has successfully authenticated with the SSO.
+        """
+        Validates a CAS ticket sent by the client for login and authenticates the user with SSO.
```

Comment:
the authentication has already happened...

Reply:
It authenticates the user with the homeserver, not with SSO. I can probably just remove the second half of the comment though.